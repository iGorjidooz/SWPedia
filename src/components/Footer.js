import React from 'react';
import { footer, footerContent, footerLink } from '../styles/FooterStyles';

const Footer = () => (
   <footer className={footer}>
      <div className={footerContent}>
         Created By <a className={footerLink} href="https://github.com/iGorjidooz">Iman Gorjidooz</a>
      </div>
   </footer>
);

export default Footer;