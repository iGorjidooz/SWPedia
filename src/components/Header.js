import React from 'react';
import { Link } from 'react-router-dom';
import * as generalStyles from '../styles/GeneralStyles';
import * as headerStyles from '../styles/HeaderStyle';

const Header = () => (
   <header className={headerStyles.header}>
      <div className={generalStyles.container}>
         <Link to="/" className={headerStyles.header__logo}>SWPedia</Link>
      </div>
   </header>
);

export default Header;