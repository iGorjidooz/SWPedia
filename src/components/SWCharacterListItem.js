import React from 'react';
import { Link } from 'react-router-dom';

const SWCharacterListItem = ({ character, listIndex }) => {
   return (
      <Link to={`/character/${character.id}`}>
         <h3>{listIndex} {character.name}</h3>
         <p>Birth Year: {character.birth_year}</p>
         <p>Gender: {character.gender}</p>
      </Link>
   );
};

export default SWCharacterListItem;