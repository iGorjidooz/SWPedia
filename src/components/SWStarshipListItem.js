import React from 'react';
import { Link } from 'react-router-dom';

const SWStarshipListItem = ({ starship, listIndex }) => {
   return (
      <Link to={`/starship/${starship.id}`}>
         <h3>{listIndex} {starship.name}</h3>
         <p>model: {starship.model}</p>
         <p>manufacturer: {starship.manufacturer}</p>
      </Link>
   );
};

export default SWStarshipListItem;