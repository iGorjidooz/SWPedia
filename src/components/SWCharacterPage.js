import React from 'react';
import { connect } from 'react-redux';
import { Link, Redirect } from 'react-router-dom';
import Header from './Header';

export const SWCharacterPage = (props) => {
   return props.character === undefined ? (
      <Redirect to="/"/>
   ) : (
      <div>
         <Header />
         <Link to="/">Back to List</Link>
         <h1>{props.character.name}</h1>
         <div>
            Profile:
            height: {props.character.height} <br/>
            mass: {props.character.mass} <br />
            Hair Color: {props.character.hair_color} <br />
            Skin color: {props.character.skin_color} <br />
            Eye Color: {props.character.eye_color} <br />
            Birth Year: {props.character.eye_color} <br />
            Gender: {props.character.gender} <br />
         </div>
      </div>
   )
};

const mapStateToProps = (state, props) => ({
   character: state.people.items.find((character) => character.id === props.match.params.id)
});

export default connect(mapStateToProps)(SWCharacterPage);