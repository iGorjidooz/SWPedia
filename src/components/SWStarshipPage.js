import React from 'react';
import { connect } from 'react-redux';
import { Link, Redirect } from 'react-router-dom';
import Header from './Header';

export const SWStarshipPage = (props) => {
   return props.starship === undefined ? (
      <Redirect to="/"/>
   ) : (
      <div>
         <Header />
         <Link to="/">Back to List</Link>
         <h1>{props.starship.name}</h1>
         <div>
            General Info: <br/>
            model: {props.starship.model} <br/>
            manufacturer: {props.starship.manufacturer} <br/>
            starship class: {props.starship.starship_class} <br/>
            Cost in credits: {props.starship.cost_in_credits} <br/>
            length: {props.starship.length}
         </div>
         <div>
            Performance: <br/>
            Max Atomsphering Speed: {props.starship.max_atmosphering_speed} <br />
            Hyperdrive Rating: {props.starship.hyperdrive_rating} <br />
            MGLT: {props.starship.MGLT} <br />
         </div>
         <div>
            Capacity: <br />
            crew: {props.starship.crew} <br/>
            passengers: {props.starship.passengers} <br />
            cargo Capacity: {props.starship.cargo_capacity} <br />
            consumables: {props.starship.consumables}
         </div>
      </div>
   )
};

const mapStateToProps = (state, props) => ({
   starship: state.starships.items.find((starship) => starship.id === props.match.params.id)
});

export default connect(mapStateToProps)(SWStarshipPage);