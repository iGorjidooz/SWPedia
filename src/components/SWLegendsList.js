import React from 'react';
import { connect } from 'react-redux';
import SWStarshipListItem from './SWStarshipListItem';
import SWCharacterListItem from './SWCharacterListItem';

export const SWLegendsList = (props) => {
   return (
      <div>
         {props.starshipItems.map((item, arrayIndex) => {
            const listIndex = arrayIndex + Math.floor(arrayIndex / 8) + 1;
            if (props.peopleItems.length > 0 && arrayIndex > 0 && (arrayIndex + 1) % 8 === 0) {
               const characterIndex = Math.floor(arrayIndex / 8);
               return (
                  <React.Fragment key={'listItemsWrapper' + arrayIndex}>
                     <SWStarshipListItem key={listIndex} starship={item} listIndex={listIndex} />
                     <SWCharacterListItem key={listIndex + 1} character={props.peopleItems[characterIndex]} listIndex={listIndex + 1} />
                  </React.Fragment>
               );
            } else {
               return (<SWStarshipListItem key={listIndex} starship={item} listIndex={listIndex} />);
            }
         })}
         {props.isFetching && <div>Loading...</div>}
      </div>
   );
};

const mapStateToProps = (state) => ({
   starshipItems: state.starships.items,
   peopleItems: state.people.items,
   isFetching: state.starships.isFetching || state.people.isFetching
});

export default connect(mapStateToProps)(SWLegendsList);