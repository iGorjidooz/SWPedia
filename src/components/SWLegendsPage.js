import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import throttle from 'lodash/throttle';
import Header from './Header';
import SWLegendsList from './SWLegendsList';
import { startFetchingStarships } from '../actions/starshipsActions';
import { startFetchingPeople } from '../actions/peopleActions';

export const SWLegendsPage = props => {
   // Scroll Event Handler Function
   // When the page scrolls to bottom it dispatches the actions to load next page of starships and people data, and unbind itself from event listener
   // throttle function from Lodash library has been used to improve performance of the scroll event handler
   const scrollEventHandler = throttle(e => {
      if(window.innerHeight + document.body.scrollTop < (document.body.offsetHeight - 20)) {
         return;
      }
      if(!props.isFetching) {
         window.removeEventListener('scroll', scrollEventHandler, { passive: true });
         props.startFetchingStarships(props.nextStarshipsPageUrl);
         props.startFetchingPeople(props.nextPeoplePageUrl);
      }
   }, 250);

   // Similar to componentDidUpdate lifecycle method
   // Looks for changes in props.isFetching and props.hasMoreStarships and binds or unbinds scrollEventHandler function to scroll event listener based on the condition
   useEffect(() => {
      props.hasMoreStarships && !props.isFetching ? window.addEventListener('scroll', scrollEventHandler, { passive: true }) : 
         window.removeEventListener('scroll', scrollEventHandler, { passive: true });
      return () => {
         window.removeEventListener('scroll', scrollEventHandler, { passive: true });
      }
   }, [props.isFetching, props.hasMoreStarships]);

   return (
      <div>
         <Header />
         <div id="legendsList">
            <SWLegendsList />
         </div>
      </div>
   );
};

const mapStateToProps = state => ({
   hasMoreStarships: !!state.starships.nextPageUrl,
   isFetching:  state.people.isFetching || state.starships.isFetching,
   nextStarshipsPageUrl: state.starships.nextPageUrl,
   nextPeoplePageUrl: state.people.nextPageUrl
});

const mapDispatchToProps = (dispatch) => ({
   startFetchingStarships: (nextStarshipsPageUrl) => dispatch(startFetchingStarships(nextStarshipsPageUrl)),
   startFetchingPeople: (nextPeoplePageUrl) => dispatch(startFetchingPeople(nextPeoplePageUrl))
 });

// Connecting the component to Redux store
export default connect(mapStateToProps, mapDispatchToProps)(SWLegendsPage);