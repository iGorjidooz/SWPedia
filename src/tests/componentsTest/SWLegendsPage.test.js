import React from 'react';
import { shallow } from 'enzyme';
import throttle from 'lodash/throttle';
import starshipsFixtures from '../fixtures/starshipsFixtures';
import peopleFixtures from '../fixtures/peopleFixtures';
import { SWLegendsPage } from '../../components/SWLegendsPage';

let wrapper, startFetchingStarships, startFetchingPeople, map;

// Forcing snapshot serializer to show the connected component name instead of generic 'ConnectFunction'
jest.mock('react-redux', () => ({
   connect: jest.fn(() => component => `Connect(${component.name})`),
}));

jest.useFakeTimers();

// jest.mock('lodash/throttle');

// throttle.mockImplementation(fn => fn);

beforeEach(() => {
   global.innerHeight = 500;
   global.document.body.scrollTop = 100;
   Object.defineProperties(global.HTMLElement.prototype, {
      offsetHeight: {
         get: function() { return parseFloat(global.getComputedStyle(this).height) || 1500; }
      },
   });

   map = {};
   window.addEventListener = jest.fn((event, cb) => {
      map[event] = cb;
   });

   startFetchingPeople = jest.fn();
   startFetchingStarships = jest.fn();
   wrapper = shallow(<SWLegendsPage
      hasMoreStarships={false}
      isFetching={false}
      nextStarshipsPageUrl={''}
      nextPeoplePageUrl={''}
      startFetchingStarships={startFetchingPeople}
      startFetchingPeople={startFetchingStarships}
   />);

   // wrapper.instance().scrollEventHandler = jest.fn();
});

test('should render SWLegendsPage correctly', () => {
   // global.document.dispatchEvent(new Event('scroll'));
   // jest.runOnlyPendingTimers();
   // expect(scrollEventHandler).toHaveBeenCalledTimes(0);
   expect(wrapper).toMatchSnapshot();
});

test('should call scrollEventHandler but not start to fetch new data', () => {
   // global.document.body.scrollTop = 1000;
   // wrapper.setProps({ hasMoreStarships: true });
   // global.document.dispatchEvent(new Event('scroll'));
   // console.log('map',map);
   // map.scroll();
   // // jest.runOnlyPendingTimers();
   // wrapper.update();
   // expect(startFetchingStarships).toHaveBeenCalledTimes(1);
})