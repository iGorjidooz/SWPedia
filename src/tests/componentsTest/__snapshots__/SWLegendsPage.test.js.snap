// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SWLegendsPage correctly 1`] = `
<div>
  <Header />
  <div
    id="legendsList"
  >
    <Connect(SWLegendsList) />
  </div>
</div>
`;
